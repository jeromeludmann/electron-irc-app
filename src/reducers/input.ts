import { Action } from 'redux'
import {
  UPDATE_INPUT_VALUE,
  UpdateInputValueAction,
  ENTER_INPUT_VALUE,
  GO_BACK_INPUT_HISTORY,
  GO_FORWARD_INPUT_HISTORY,
  EnterInputValueAction,
} from '@app/actions/ui'

export type InputState = Readonly<{
  value: string
  dirtyValue: string
  history: Readonly<{
    values: string[]
    index: number
  }>
}>

type InputReducer = (input: InputState, action: Action) => InputState

export const inputInitialState: InputState = {
  value: '',
  dirtyValue: '',
  history: { values: [], index: 0 },
}

const beginOfHistory = (input: InputState) => {
  return input.history.index === 0
}

const endOfHistory = (input: InputState) => {
  return input.history.index === input.history.values.length
}

const handlers: { [action: string]: InputReducer } = {
  [UPDATE_INPUT_VALUE]: (input, action: UpdateInputValueAction) => ({
    ...input,
    value: action.payload.value,
    dirtyValue: endOfHistory(input) ? action.payload.value : input.dirtyValue,
  }),

  [ENTER_INPUT_VALUE]: (input, action: EnterInputValueAction) => {
    const historyValues = [...input.history.values, action.payload.value]
    return {
      value: '',
      dirtyValue: '',
      history: { values: historyValues, index: historyValues.length },
    }
  },

  [GO_BACK_INPUT_HISTORY]: input =>
    beginOfHistory(input)
      ? input
      : {
          ...input,
          value: input.history.values[input.history.index - 1],
          history: { ...input.history, index: input.history.index - 1 },
        },

  [GO_FORWARD_INPUT_HISTORY]: input =>
    endOfHistory(input)
      ? input
      : {
          ...input,
          value:
            input.history.values[input.history.index + 1] || input.dirtyValue,
          history: { ...input.history, index: input.history.index + 1 },
        },
}

export const reduceInput: InputReducer = (input = inputInitialState, action) =>
  action.type in handlers ? handlers[action.type](input, action) : input
