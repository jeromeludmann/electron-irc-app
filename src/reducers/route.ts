import { Action } from 'redux'
import { Route, BufferKey, isStatus } from '@app/Route'
import {
  SWITCH_WINDOW,
  SwitchWindowAction,
  CLOSE_WINDOW,
  CloseWindowAction,
} from '@app/actions/ui'
import { ReceiveJoinAction, RECEIVE_JOIN } from '@app/actions/msgIncoming'
import { RootState } from '@app/reducers'

export type RouteState = Readonly<Route>

export const routeInitialState: RouteState = {
  serverKey: '',
  bufferKey: BufferKey.STATUS,
}

type RouteReducer = (
  state: RouteState,
  action: Action,
  extraStates: { root: RootState },
) => RouteState

const caseReducers: { [action: string]: RouteReducer } = {
  [RECEIVE_JOIN]: (route, action: ReceiveJoinAction, extraStates) =>
    action.payload.user.nick ===
    extraStates.root.servers[action.route.serverKey].user.nick
      ? action.route
      : route,

  [CLOSE_WINDOW]: (route, _: CloseWindowAction, extraStates) => {
    if (isStatus(route.bufferKey)) {
      const keys = Object.keys(extraStates.root.servers)
      return keys.length > 1
        ? {
            serverKey: keys.filter(key => key !== route.serverKey)[0],
            bufferKey: BufferKey.STATUS,
          }
        : route
    }

    return { ...route, bufferKey: BufferKey.STATUS }
  },

  // TODO use extraStates and check if the given route exists
  [SWITCH_WINDOW]: (_, action: SwitchWindowAction) => action.route,
}

export const reduceRoute: RouteReducer = (
  routeState = routeInitialState,
  action,
  extraStates,
) =>
  action.type in caseReducers
    ? caseReducers[action.type](routeState, action, extraStates)
    : routeState
