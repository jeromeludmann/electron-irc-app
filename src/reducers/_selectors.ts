import { createSelector } from 'reselect'
import { RootState } from '@app/reducers'
import { RouteState } from '@app/reducers/route'
import { ServerState } from '@app/reducers/server'

export const selectServers = ({
  servers,
}: RootState): {
  [key: string]: ServerState
} => servers

export const selectRoute = ({ route }: RootState): RouteState => route

export const selectServer = createSelector(
  selectServers,
  selectRoute,
  (servers, { serverKey }) => servers[serverKey],
)

export const selectUser = createSelector(selectServer, server => server.user)

export const selectBuffer = createSelector(
  selectServer,
  selectRoute,
  (server, active) => server.buffers[active.bufferKey],
)

export const selectMessages = createSelector(
  selectBuffer,
  channel => channel.messages,
)

export const selectInput = createSelector(
  selectBuffer,
  channel => channel.input,
)

export const selectValue = createSelector(selectInput, input => input.value)
