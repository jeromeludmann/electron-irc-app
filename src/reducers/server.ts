import { RouteState } from '@app/reducers/route'
import {
  BufferRouterState,
  bufferRouterInitialState,
  reduceBufferRouter,
} from '@app/reducers/bufferRouter'
import { RoutedAction } from '@app/Route'
import {
  RECEIVE_RPL_MYINFO,
  ReceiveReplyMyInfoAction,
  RECEIVE_NICK,
  ReceiveNickAction,
  RECEIVE_PONG_FROM_SERVER,
  ReceivePongFromServerAction,
} from '@app/actions/msgIncoming'

type ServerPartialState = Readonly<{
  name: string
  user: Readonly<{
    nick: string
    user: string
    real: string
  }>
  lag: number
  // TODO refactor to another nested reducer
  modes: Readonly<{
    user: string[]
    available: Readonly<{
      channel: string[]
      user: string[]
    }>
  }>
}>

export type ServerState = Readonly<{
  buffers: BufferRouterState
}> &
  ServerPartialState

type ServerReducer<S = ServerState> = (
  server: S,
  action: RoutedAction,
  extraStates: { route: RouteState },
) => S

export const serverInitialState = {
  name: '<unknown>',
  user: {
    nick: 'default_nick',
    user: 'default_user',
    real: 'default_name',
  },
  lag: 0,
  modes: {
    user: [],
    available: { channel: [], user: [] },
  },
  buffers: bufferRouterInitialState,
}

const handlers: { [action: string]: ServerReducer<ServerPartialState> } = {
  [RECEIVE_NICK]: (server, action: ReceiveNickAction) => ({
    ...server,
    user:
      action.payload.user.nick === server.user.nick
        ? { ...server.user, nick: action.payload.nick }
        : server.user,
  }),

  [RECEIVE_PONG_FROM_SERVER]: (
    server,
    action: ReceivePongFromServerAction,
  ) => ({
    ...server,
    lag: action.payload.lag,
  }),

  [RECEIVE_RPL_MYINFO]: (server, action: ReceiveReplyMyInfoAction) => ({
    ...server,
    name: action.payload.serverName,
    modes: {
      ...server.modes,
      available: {
        channel: action.payload.availableChannelModes,
        user: action.payload.availableUserModes,
      },
    },
  }),
}

export const reduceServer: ServerReducer = (
  server = serverInitialState,
  action,
  extraStates,
) => ({
  ...(action.type in handlers
    ? handlers[action.type](server, action, extraStates)
    : server),
  buffers: reduceBufferRouter(server.buffers, action as RoutedAction, {
    ...extraStates,
    server,
  }),
})
