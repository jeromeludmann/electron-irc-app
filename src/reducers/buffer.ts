import {
  MessagesState,
  messagesInitialState,
  reduceMessages,
} from '@app/reducers/messages'
import { InputState, inputInitialState, reduceInput } from '@app/reducers/input'
import { RoutedAction } from '@app/Route'
import { RouteState } from '@app/reducers/route'
import { RECEIVE_JOIN, ReceiveJoinAction } from '@app/actions/msgIncoming'
import { SWITCH_WINDOW } from '@app/actions/ui'
import { ServerState } from '@app/reducers/server'

type BufferPartialState = Readonly<{
  activity: boolean
}>

export type BufferState = Readonly<{
  input: InputState
  messages: MessagesState
}> &
  BufferPartialState

type BufferReducer<S = BufferState> = (
  buffer: S,
  action: RoutedAction,
  extraStates: {
    route: RouteState
    server: ServerState
  },
) => S

export const bufferInitialState = {
  activity: false,
  input: inputInitialState,
  messages: messagesInitialState,
}

const handlers: { [action: string]: BufferReducer<BufferPartialState> } = {
  [RECEIVE_JOIN]: (_, action: ReceiveJoinAction, extraStates) => ({
    activity: action.payload.user.nick !== extraStates.server.user.nick,
  }),

  [SWITCH_WINDOW]: () => ({ activity: false }),
}

export const reduceBuffer: BufferReducer = (
  buffer = bufferInitialState,
  action,
  extraStates,
): BufferState => ({
  ...(handlers.hasOwnProperty(action.type)
    ? handlers[action.type](buffer, action, extraStates)
    : buffer),
  input: reduceInput(buffer.input, action),
  messages: reduceMessages(buffer.messages, action, extraStates),
})
