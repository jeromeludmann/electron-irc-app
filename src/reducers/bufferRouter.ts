import {
  BufferState,
  bufferInitialState,
  reduceBuffer,
} from '@app/reducers/buffer'
import {
  BufferKey,
  RoutedAction,
  isRaw,
  isPrivate,
  isChannel,
} from '@app/Route'
import { RouteState } from '@app/reducers/route'
import { CLOSE_WINDOW, CloseWindowAction } from '@app/actions/ui'
import { RECEIVE_JOIN, RECEIVE_PRIVMSG } from '@app/actions/msgIncoming'
import { ServerState } from '@app/reducers/server'

export type BufferRouterState = Readonly<{
  [key: string]: BufferState
}>

type BufferRouterReducer = (
  buffers: BufferRouterState,
  action: RoutedAction,
  extraStates: { route: RouteState; server: ServerState },
) => BufferRouterState

export const bufferRouterInitialState: BufferRouterState = {
  [BufferKey.RAW]: bufferInitialState,
  [BufferKey.STATUS]: bufferInitialState,
}

const routeHandlers: { [buffer: string]: BufferRouterReducer } = {
  [BufferKey.NONE]: buffers => buffers,

  [BufferKey.ACTIVE]: (buffers, action, extraStates) => {
    const key = extraStates.route.bufferKey
    return {
      ...buffers,
      [key]: reduceBuffer(buffers[key], action, extraStates),
    }
  },

  [BufferKey.ALL]: (buffers, action, extraStates) => {
    const toBroadcast: { [key: string]: BufferState } = {}

    Object.keys(buffers).forEach(key => {
      if (!isRaw(key)) {
        toBroadcast[key] = reduceBuffer(buffers[key], action, extraStates)
      }
    })

    return { ...buffers, ...toBroadcast }
  },
}

const handlers: { [action: string]: BufferRouterReducer } = {
  [CLOSE_WINDOW]: (buffers, action: CloseWindowAction) => {
    const updatedBuffers = { ...buffers }

    if (
      isChannel(action.route.bufferKey) ||
      isPrivate(action.route.bufferKey)
    ) {
      delete updatedBuffers[action.route.bufferKey]
      return updatedBuffers
    }

    // Remove all server-related buffers while closing status window
    Object.keys(buffers).forEach(buffer => {
      if (isChannel(buffer) || isPrivate(buffer)) {
        delete updatedBuffers[buffer]
      }
    })

    return updatedBuffers
  },
}

export const reduceBufferRouter: BufferRouterReducer = (
  buffers = bufferRouterInitialState,
  action,
  extraStates,
) => {
  const { bufferKey } = action.route

  if (bufferKey in routeHandlers) {
    return routeHandlers[bufferKey](buffers, action, extraStates)
  }

  if (action.type in handlers) {
    return handlers[action.type](buffers, action, extraStates)
  }

  if (
    bufferKey in buffers ||
    [RECEIVE_PRIVMSG, RECEIVE_JOIN].indexOf(action.type) > -1
  ) {
    return {
      ...buffers,
      // TODO check the route type and use appropriate reducer?
      [bufferKey]: reduceBuffer(buffers[bufferKey], action, extraStates),
    }
  }

  return buffers
}
