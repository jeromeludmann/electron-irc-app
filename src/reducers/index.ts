import { Reducer } from 'redux'
import { RoutedAction, isChannel, isPrivate } from '@app/Route'
import { RouteState, routeInitialState, reduceRoute } from '@app/reducers/route'
import { ServerState, reduceServer } from '@app/reducers/server'
import { CLOSE_WINDOW } from '@app/actions/ui'
import { CONNECT_SERVER } from '@app/actions/socket'

type RootPartialState = Readonly<{
  servers: Readonly<{ [key: string]: ServerState }>
}>

export type RootState = Readonly<{
  route: RouteState
}> &
  RootPartialState

type RootReducer<S = RootState> = (
  root: S,
  action: RoutedAction,
  extraStates: { route: RouteState },
) => S

export const rootInitialState = {
  servers: {},
  route: routeInitialState,
}

const routeActionToServer: RootReducer<
  Readonly<{ [key: string]: ServerState }>
> = (servers, action, extraStates) => {
  const serverKeyFound =
    action.route !== undefined && action.route.serverKey in servers

  return serverKeyFound || action.type === CONNECT_SERVER
    ? {
        ...servers,
        [action.route.serverKey]: reduceServer(
          servers[action.route.serverKey],
          action,
          { route: extraStates.route },
        ),
      }
    : servers
}

const handlers: { [action: string]: RootReducer<RootPartialState> } = {
  [CLOSE_WINDOW]: (root, action, extraStates) => {
    const thereIsOnlyOneServer = Object.keys(root.servers).length <= 1
    const isChannelOrPrivate =
      isChannel(action.route.bufferKey) || isPrivate(action.route.bufferKey)

    if (thereIsOnlyOneServer || isChannelOrPrivate) {
      return { servers: routeActionToServer(root.servers, action, extraStates) }
    }

    const updatedServers = { ...root.servers }
    delete updatedServers[action.route.serverKey]
    return { servers: updatedServers }
  },
}

export const reduceRoot: Reducer<RootState, RoutedAction> = (
  root = rootInitialState,
  action,
) => {
  if (action.route === undefined) {
    return root
  }

  return {
    ...(action.type in handlers
      ? handlers[action.type](root, action, { route: root.route })
      : {
          servers: routeActionToServer(root.servers, action, {
            route: root.route,
          }),
        }),
    route: reduceRoute(root.route, action, { root }),
  }
}
