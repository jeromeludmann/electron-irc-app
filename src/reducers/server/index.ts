import {
  ServerModesState,
  serverModesInitialState,
  reduceServerModes,
} from "@app/reducers/server/modes";
import { RouteState } from "@app/reducers/route";
import {
  ChannelsState,
  channelsInitialState,
  reduceChannels,
} from "@app/reducers/server/channels";
import { RoutedAction } from "@app/Route";
import {
  serverNameInitialState,
  reduceServerName,
} from "@app/reducers/server/name";
import {
  reduceAvailableServerModes,
  AvailableServerModesState,
  availableServerModesInitialState,
} from "@app/reducers/server/availableModes";
import { ServerNameState } from "@app/reducers/server/name";
import {
  UserState,
  userInitialState,
  reduceUser,
} from "@app/reducers/server/user";
import {
  reduceLag,
  ServerLagState,
  serverLagInitialState,
} from "@app/reducers/server/lag";
import { Action } from "redux";

export type ServerState = Readonly<{
  name: ServerNameState;
  user: UserState;
  availableModes: AvailableServerModesState;
  modes: ServerModesState;
  channels: ChannelsState;
  lag: ServerLagState;
}>;

export const serverInitialState: ServerState = {
  name: serverNameInitialState,
  user: userInitialState,
  availableModes: availableServerModesInitialState,
  modes: serverModesInitialState,
  channels: channelsInitialState,
  lag: serverLagInitialState,
};

export const reduceServer = (
  server = serverInitialState,
  action: Action,
  extraStates: { route: RouteState },
): ServerState => ({
  name: reduceServerName(server.name, action),
  user: reduceUser(server.user, action),
  availableModes: reduceAvailableServerModes(server.availableModes, action),
  modes: reduceServerModes(server.modes),
  channels: reduceChannels(server.channels, action as RoutedAction, {
    ...extraStates,
    user: server.user,
  }),
  lag: reduceLag(server.lag, action),
});
