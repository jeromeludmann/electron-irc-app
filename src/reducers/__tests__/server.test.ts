import { reduceServer } from "@app/reducers/server";
import { messageReceivers } from "@app/actions/msgIncoming";

describe("reduce server", () => {
  const route = { serverKey: "server1", channelKey: "#channel" };
  const extraStates = { route };

  expect(
    reduceServer(undefined, { type: "", route }, extraStates),
  ).toMatchSnapshot();

  describe("name", () => {
    it("without action", () => {
      expect(
        reduceServer(undefined, { type: "", route }, extraStates),
      ).toMatchSnapshot();
    });

    it("with RPL_MYINFO received", () => {
      expect(
        reduceServer(
          undefined,
          messageReceivers["004"]("server1", "server", [
            "",
            "serverName",
            "version",
            "userModes",
            "channelModes",
          ]),
          extraStates,
        ),
      ).toMatchSnapshot();
    });
  });
});
