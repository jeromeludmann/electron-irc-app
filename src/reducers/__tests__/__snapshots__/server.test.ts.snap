// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "buffers": Object {
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "default_nick",
    "real": "default_name",
    "user": "default_user",
  },
}
`;

exports[`reduce server name with RPL_MYINFO received 1`] = `
Object {
  "buffers": Object {
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": true,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [
        "c",
        "h",
        "a",
        "n",
        "n",
        "e",
        "l",
        "M",
        "o",
        "d",
        "e",
        "s",
      ],
      "user": Array [
        "u",
        "s",
        "e",
        "r",
        "M",
        "o",
        "d",
        "e",
        "s",
      ],
    },
    "user": Array [],
  },
  "name": "serverName",
  "user": Object {
    "nick": "default_nick",
    "real": "default_name",
    "user": "default_user",
  },
}
`;

exports[`reduce server name without action 1`] = `
Object {
  "buffers": Object {
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "default_nick",
    "real": "default_name",
    "user": "default_user",
  },
}
`;
